sshpiperd/upstream/database/handler.go

```diff
@@ -2,86 +2,313 @@ package database
+	MappedUsername        string
+	FromType              authMapType
+	FromPassword          string
+	FromPrivateKey        downstreamPrivateKey
+	FromAuthorizedKeys    []downstreamAuthorizedKey
+	FromAllowAnyPublicKey bool
+	ToType                authMapType
+	ToPassword            string
+	ToPrivateKey          upstreamPrivateKey
+	ToAuthorizedKeys      []upstreamAuthorizedKey
+	NoPassthrough         bool
+	KnownHosts            string
+	KnownHostsData        string
+	IgnoreHostkey         bool
+}
+
+func (p *plugin) loadPipeFromDB(conn ssh.ConnMetadata) (pipeConfig, error) {
 
 	user := conn.User()
 	d, err := lookupDownstreamWithFallback(p.db, user)
 
 	if err != nil {
-		return nil, nil, err
+		return pipeConfig{}, err
 	}
 
-	addr := d.Upstream.Server.Address
-	upuser := d.Upstream.Username
-
-	if upuser == "" {
-		upuser = d.Username
+	pipe := pipeConfig{
+		Username:              user,
+		UpstreamHost:          d.Upstream.Server.Address,
+		MappedUsername:        d.Upstream.Username,
+		FromType:              d.AuthMapType,
+		FromPassword:          d.Password,
+		FromAuthorizedKeys:    d.AuthorizedKeys,
+		FromAllowAnyPublicKey: d.AllowAnyPublicKey,
+		ToType:                d.Upstream.AuthMapType,
+		ToPassword:            d.Upstream.Password,
+		ToPrivateKey:          d.Upstream.PrivateKey,
+		ToAuthorizedKeys:      d.Upstream.AuthorizedKeys,
+		NoPassthrough:         d.NoPassthrough,
+		KnownHosts:            d.Upstream.KnownHosts,
+		IgnoreHostkey:         d.Upstream.Server.IgnoreHostKey,
 	}
 
-	logger.Printf("mapping user [%v] to [%v@%v]", user, upuser, addr)
+	return pipe, nil
+}
 
-	c, err := upstreamprovider.DialForSSH(addr)
+func (p *plugin) GetBytes(inputString string) ([]byte, error) {
 
-	if err != nil {
-		return nil, nil, err
+	if inputString != "" {
+		return []byte(inputString), nil
 	}
 
+	return nil, nil
+}
+
+func (p *plugin) createAuthPipe(pipe pipeConfig, conn ssh.ConnMetadata, challengeContext ssh.AdditionalChallengeContext) (*ssh.AuthPipe, error) {
+
 	hostKeyCallback := ssh.InsecureIgnoreHostKey()
+	if !pipe.IgnoreHostkey {
 
-	if !d.Upstream.Server.IgnoreHostKey {
+		data, err := p.GetBytes(pipe.KnownHosts)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(data) == 0 {
+			return nil, fmt.Errorf("no known hosts spicified")
+		}
 
-		key, _, _, _, err := ssh.ParseAuthorizedKey([]byte(d.Upstream.Server.HostKey.Key.Data))
+		hostKeyCallback, err = knownhosts.NewFromReader(bytes.NewReader(data))
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
+	}
+
+	to := func(key ssh.PublicKey) (ssh.AuthPipeType, ssh.AuthMethod, error) {
+
+		switch pipe.ToType {
+		case authMapTypeNone:
+			return ssh.AuthPipeTypeNone, nil, nil
+
+		case authMapTypePassword:
+			toPass, err := p.GetBytes(pipe.ToPassword)
+
+			if err != nil {
+				return ssh.AuthPipeTypeDiscard, nil, err
+			}
 
-		hostKeyCallback = ssh.FixedHostKey(key)
+			return ssh.AuthPipeTypeMap, ssh.Password(string(toPass)), nil
```

Comment:
why tobyte then to string again?

Reply:
removed also this unnecessary conversion